### ISSUE TITLE
`System.InvalidOperationException` when getting a `404 NOT FOUND` response from an endpoint

### ISSUE THREAD

**Issue by candoumbe** [FastEndpoints/FastEndpoints#422]
I'm just starting using this library to create set of microservices using the vertical slice feature approach.
The endpoint under test

```csharp

/// <summary>
/// Gets an appointment by its id
/// </summary>
public class Endpoint : Endpoint<AppointmentId, Browsable<AppointmentInfo>>
{
    private readonly IUnitOfWorkFactory _unitOfWorkFactory;
    private readonly LinkGenerator _linkGenerator;


    /// <inheritdoc/>
    public override void Configure()
    {
        Get("/appointments/{@id}", x => new { Id = x.Value });
        Head("/appointments/{@id}", x => new { Id = x.Value });
        AllowAnonymous();
    }

    /// <inheritdoc/>
    public override async Task HandleAsync(AppointmentId req, CancellationToken ct)
    {
        using IUnitOfWork unitOfWork = _unitOfWorkFactory.NewUnitOfWork();
        Option<AppointmentInfo> mayBeAppointment = await unitOfWork.Repository<Appointment>()
                                                                   .SingleOrDefault(selector: appointment => new AppointmentInfo
                                                                   {
                                                                       Id = appointment.Id,
                                                                       Subject = appointment.Subject,
                                                                       Location = appointment.Location,
                                                                       EndDate = appointment.EndDate,
                                                                       StartDate = appointment.StartDate,
                                                                       Attendees = appointment.Attendees.Select(attendee => new AttendeeInfo
                                                                       {
                                                                           Id = attendee.Id,
                                                                           Email = attendee.Email,
                                                                           Name = attendee.Name,
                                                                           PhoneNumber = attendee.PhoneNumber,
                                                                       }),
                                                                   },
                                                                                    predicate: (Appointment x) => x.Id == req,
                                                                                    cancellationToken: ct).ConfigureAwait(false);

        await mayBeAppointment.Match(
            some: async appointment =>
            {
                Browsable<AppointmentInfo> browsable = new()
                {
                    Resource = appointment,
                    Links = new[]
                    {
                        new Link
                        {
                            Href = _linkGenerator.GetUriByName(HttpContext, nameof(Endpoint), new { Id = req }),
                            Method = "GET",
                            Relations = new [] { LinkRelation.Self }.ToHashSet()
                        },
                    }
                };
                await SendOkAsync(browsable, ct).ConfigureAwait(false);
            },
            none: async () => await SendNotFoundAsync(ct).ConfigureAwait(false))
                              .ConfigureAwait(false);
    }
}
```

The test basically checks that I indeed get a 404 when I hit the endpoint with an unknown id
```csharp
    public class EndpointShould : IClassFixture<AgendaWebApplicationFactory>
    {
        private readonly HttpClient _client;
        private readonly ITestOutputHelper _outputHelper;
        private readonly AgendaWebApplicationFactory _applicationFactory;

        [Fact]
        public async Task Returns_NotFound_when_Id_does_not_exist()
        {
            // Arrange
            AppointmentId appointmentId = AppointmentId.New();

            // Act
            (HttpResponseMessage response, AppointmentInfo _) = await _client.GETAsync<Endpoint, AppointmentId, AppointmentInfo>(appointmentId);

            // Assert
            response.StatusCode.Should().Be(System.Net.HttpStatusCode.NotFound);
        }
    }
```

I get the following error as soon as I run the test
```log
 Message: 
System.InvalidOperationException : Unable to deserialize the response body as [Agenda.API.Resources.v1.Appointments.AppointmentInfo]. Reason: [NotFound] 

  Arborescence des appels de procédure: 
HttpClientExtensions.Send[TRequest,TResponse](HttpClient client, HttpMethod method, String requestUri, TRequest request, Nullable`1 isFormContent)
```

This seems to be due to the [HttpClientExtensions.Send](https://github.com/FastEndpoints/FastEndpoints/blob/a1e3f60d21480d60f5d5daee3ec78d5917b7c511/Src/Library/Testing/HttpClientExtensions.cs#L228) method which always deserializes the json for endpoints that are supposed to returns a response, no matter what the HTTP status code is.

A fix could be something like the following

```csharp
        if (typeof(TResponse) == Types.EmptyResponse)
            return new(rsp, default);

        TResponse? res;
       
       if(CanDeserializeContent(rsp)) // <-- This method would check that there is a content to deserialize based on status code, content directly etc....
      {
        try
        {
            res = await rsp.Content.ReadFromJsonAsync<TResponse>(SerOpts.Options);
        }
        catch (JsonException)
        {
            var reason = $"[{rsp.StatusCode}] {await rsp.Content.ReadAsStringAsync()}";
            throw new InvalidOperationException(
                $"Unable to deserialize the response body as [{typeof(TResponse).FullName}]. Reason: {reason}");
        }
   }
        return new(rsp, res);
    }


   static bool CanDeserialize(HttpResponseMessage response) => response.StatusCode != 404
```

I can make a PR if it's okay for you

**Comment by candoumbe** [comment]
I don't know if this issue is related to #267

**Comment by dj-nitehawk** [comment]
it's not possible to change the behavior of the [Send()](https://github.com/FastEndpoints/FastEndpoints/blob/main/Src/Library/Testing/HttpClientExtensions.cs#L203) method as it would break a lot of existing tests written by users.

however, starting from `v5.8.1.8-beta`, you can do the following to assert the failure path:
https://github.com/FastEndpoints/FastEndpoints/blob/cea5b3526c59dd8836e7b462e1cc025116d71f44/Tests/IntegrationTests/FastEndpoints.IntegrationTests/WebTests/AdminTests.cs#L74-L93

**Comment by candoumbe** [comment]
Unfortunately that's not the expected behavior when posting or getting a response from the endpoint 😔
The behaviour could be change (resulting in a breaking change I understand that) but when retrieving an response, I wouldn't expect the deserialization to fail in that case because :
it forces a developer to write differently a test for the happy path (no `try ... catch`) and others that would throw. IMHO there should be not difference in the way test are written, only expectations should differ : status codes are not the same and so are the dto.

Thanks anyway for taking the time to answer to my question and thanks for this wonderful piece of art

**Comment by dj-nitehawk** [comment]
yeah i understand. but a wrong desicion was made earlier on with the extension methods.
what you could do is simply copy the code and write your own set of extensions for the http client.

**Comment by dj-nitehawk** [comment]
@candoumbe

decided to do a breaking change and finally sort this out:
https://github.com/FastEndpoints/FastEndpoints/blob/main/Src/Library/changelog.md#breaking-changes

available in `v5.8.1.10-beta`

**Comment by candoumbe** [comment]
> @candoumbe
> 
> decided to do a breaking change and finally sort this out: https://github.com/FastEndpoints/FastEndpoints/blob/main/Src/Library/changelog.md#breaking-changes
> 
> available in `v5.8.1.10-beta`

Nice !

Is it intentional to not go for a v`6.x` which advertise the breaking changes ?

**Comment by dj-nitehawk** [comment]
final release will be v6.0 for sure.
